src/proof_rules/calculus/soundness_checker.rs

```diff
@@ -1,121 +1,310 @@
-use std::{collections::HashMap, fmt, ops::DerefMut, vec};
+use std::{
+    collections::HashMap,
+    fmt,
+    ops::{BitAnd, BitOr, DerefMut},
+    vec,
+};
+
+use itertools::Itertools;
 
 use crate::{
-    ast::{Block, DeclKind, Direction, ProcDecl, Span, Stmt, StmtKind},
+    ast::{Block, DeclKind, Direction, Label, ProcDecl, Span, Stmt, StmtKind},
     driver::{
         front::{Module, SourceUnit},
         item::SourceUnitName,
     },
     intrinsic::annotations::{AnnotationKind, Calculus},
-    proof_rules::{get_calculus, infer_fixpoint_semantics_kind},
+    proof_rules::{get_proc_calculus, infer_fixpoint_semantics_kind},
     tyctx::TyCtx,
 };
 
 #[derive(Debug, Clone)]
-pub struct ApproximationListEntry {
+pub enum StmtKindName {
+    Stmt,
+    Loop,
+}
+
+impl fmt::Display for StmtKindName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            StmtKindName::Stmt => write!(f, "statement"),
+            StmtKindName::Loop => write!(f, "loop"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ApproximationRecord {
     pub span: Span,
-    pub is_loop: bool,
+    pub stmt_kind: StmtKindName,
     pub kind: ApproximationKind,
 }
-pub type ApproximationList = Vec<ApproximationListEntry>;
+#[derive(Debug, Clone, Default)]
+pub struct ApproximationList(pub Vec<ApproximationRecord>);
 
-// pub type SoundnessBlame = (Soundness, ApproximationList);
+impl IntoIterator for ApproximationList {
+    type Item = ApproximationRecord;
+    type IntoIter = std::vec::IntoIter<ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
 
-#[derive(Debug, Clone, Default)]
-pub struct SoundnessBlame {
-    pub soundness: Soundness,
-    pub approximations: ApproximationList,
-    pub calculus: Option<Calculus>,
+impl<'a> IntoIterator for &'a ApproximationList {
+    type Item = &'a ApproximationRecord;
+    type IntoIter = std::slice::Iter<'a, ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
 }
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
-pub enum ApproximationKind {
-    #[default]
-    Exact,
-    Under,
-    Over,
-    Unknown,
+impl<'a> IntoIterator for &'a mut ApproximationList {
+    type Item = &'a mut ApproximationRecord;
+    type IntoIter = std::slice::IterMut<'a, ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter_mut()
+    }
 }
 
-impl ApproximationKind {
-    pub fn supremum(self, other: ApproximationKind) -> ApproximationKind {
-        match self.partial_cmp(&other) {
-            Some(std::cmp::Ordering::Equal) => self,
-            Some(std::cmp::Ordering::Less) => other,
-            Some(std::cmp::Ordering::Greater) => self,
-            None => ApproximationKind::Exact, // If they are incomparable, return Exact (which is the top element)
+impl ApproximationList {
+    pub fn infimum(&self) -> ApproximationKind {
```

Comment:
I suggest adding Rustdoc here, and also a bunch of [Documentation tests](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html) to explain this, esp. with the list of zero items, one item, and two/three.